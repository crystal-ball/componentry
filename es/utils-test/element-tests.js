/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { mount } from 'enzyme';
import dt from './dt';
/**
 * Library default component requirment test suite.
 *
 * IMPORTANT NOTES
 * - Enzyme `find()` will find **EVERY** instance of a prop, this makes it difficult
 *   to assert on find lengths, because it will be variable for components with
 *   HOCs which is why we use `toBeTruthy` to only check that some instance of the
 *   expected element attr exists
 */

export default (function (TestComponent) {
  var enzymeOptions = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  /*
   * Test that any html tag or component can be passed through the 'as' prop to
   * override the parent element the component renders
   */
  test('should render as specified html element or component', function () {
    // Create a component to validate that the TestComponent returns.
    var TestAs = function TestAs() {
      return /*#__PURE__*/React.createElement("section", null);
    };

    var wrapper = mount( /*#__PURE__*/React.createElement(TestComponent, {
      as: TestAs
    }), enzymeOptions);
    expect(wrapper.find(TestAs).length).toBeTruthy();
  });
  /*
   * All components should spread passed props on the parent container for max
   * customization
   */

  test('should pass through classnames and attributes', function () {
    var wrapper = mount( /*#__PURE__*/React.createElement(TestComponent, {
      className: "test-custom",
      "data-test": "test-custom"
    }), enzymeOptions);
    expect(wrapper.find('.test-custom').length).toBeTruthy();
    expect(wrapper.find(dt('test-custom')).length).toBeTruthy();
  });
});