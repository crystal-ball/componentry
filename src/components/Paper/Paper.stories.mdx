import { ArgsTable, Meta, Story } from '@storybook/addon-docs'

import { Paper } from './Paper'
import { Text } from '../Text/Text'

<Meta title='Components/Paper' component={Paper} />

# Paper

Paper is a layout component for creating containers.

<Story name='Paper'>
  <Paper p={2}>
    <Text>
      Paper is used for creating containers with consistent border, border-radius and
      box-shadow styles.
    </Text>
  </Paper>
</Story>

<ArgsTable of={Paper} />
