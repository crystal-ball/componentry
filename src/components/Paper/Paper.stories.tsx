import type { Meta, StoryObj } from '@storybook/react'

import { Paper } from './Paper'

const meta: Meta<typeof Paper> = {
  component: Paper,
}

export default meta
type Story = StoryObj<typeof Paper>

export const Primary: Story = {
  args: {
    p: 2,
    children: 'Hello World',
  },
}
