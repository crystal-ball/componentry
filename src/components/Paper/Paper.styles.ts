import { type CSSProperties } from 'react'
import { Theme } from '../../theme/theme'

//                                        <Paper /> styles
// -------------------------------------------------------

export const paperStyles = (theme: Theme): PaperStyles => ({
  // BASE
  '.C9Y-Paper-base': {
    borderRadius: theme.borderRadius.DEFAULT,
  },

  // VARIANTS
  '.C9Y-Paper-flat': {
    border: theme.border.DEFAULT,
  },
})

export interface PaperStyles {
  /** Base class applied to all variants for shared structural styles */
  '.C9Y-Paper-base': CSSProperties
  /** Variant class applied when `variant="flat"` */
  '.C9Y-Paper-flat': CSSProperties
}
