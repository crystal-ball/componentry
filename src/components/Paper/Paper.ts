import { forwardRef } from 'react'
import { createElement } from '../../utils/create-element'
import { MergeTypes, Resolve } from '../../utils/types'
import { ElementTypeProps, UtilityProps } from '../../utils/utility-props'
import { useThemeProps } from '../Provider/Provider'

/** Module augmentation interface for overriding component props' types */
export interface PaperPropsOverrides {}

export interface PaperPropsDefaults {
  /** Display variant */
  variant?: 'flat'
}

export type PaperProps<As extends React.ElementType = 'div'> = Resolve<
  MergeTypes<PaperPropsDefaults, PaperPropsOverrides> & { as?: As } & UtilityProps
> &
  ElementTypeProps<As>

/**
 * Paper provides containers for custom elements.
 * @example
 * ```tsx
 * <Paper variant="flat">
 *   ...
 * </Paper>
 * ```
 * @see [üìù Paper](https://componentry.design/docs/components/paper)
 */
export interface Paper {
  <As extends React.ElementType = 'div'>(props: PaperProps<As>): React.ReactElement
  displayName?: string
}

export const Paper = forwardRef<HTMLElement, PaperProps>((props, ref) => {
  const { variant = 'flat', ...rest } = {
    ...useThemeProps('Paper'),
    ...props,
  }

  return createElement({
    ref,
    componentClassName: ['C9Y-Paper-base', `C9Y-Paper-${variant}`],
    ...rest,
  })
}) as Paper
Paper.displayName = 'Paper'
