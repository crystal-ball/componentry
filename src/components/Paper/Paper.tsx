import { type ComponentPropsWithRef, forwardRef } from 'react'
import { element } from '../../utils/element-creator'
import { MergeTypes, Resolve } from '../../utils/types'
import { UtilityProps } from '../../utils/utility-classes'
import { useThemeProps } from '../Provider/Provider'

// Module augmentation interface for overriding component props' types
export interface PaperPropsOverrides {}

export interface PaperPropsDefaults {
  variant?: 'flat'
}

export type PaperProps = Resolve<MergeTypes<PaperPropsDefaults, PaperPropsOverrides>> &
  UtilityProps &
  ComponentPropsWithRef<'div'>

// ‚ú® Nice display type for IntelliSense
export interface Paper {
  (props: PaperProps): React.ReactElement | null
  displayName?: string
}

/**
 * #### [üìù Paper](https://componentry.design/docs/components/paper)
 *
 * Paper provides containers for custom elements.
 * @example
 * ```tsx
 * <Paper variant="flat">
 *   ...
 * </Paper>
 * ```
 */
export const Paper: Paper = forwardRef((props, ref) => {
  const { variant = 'flat', ...rest } = {
    ...useThemeProps('Paper'),
    ...props,
  }

  return element({
    ref,
    componentCx: ['C9Y-Paper-base', `C9Y-Paper-${variant}`],
    ...rest,
  })
})
Paper.displayName = 'Paper'
