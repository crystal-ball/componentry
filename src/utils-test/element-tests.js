/* eslint-disable import/no-extraneous-dependencies */
import React from 'react'
import { mount } from 'enzyme'

import dt from './dt'

/**
 * Library default component requirment test suite.
 *
 * IMPORTANT NOTES
 * - Enzyme `find()` will find **EVERY** instance of a prop, this makes it difficult
 *   to assert on find lengths, because it will be variable for components with
 *   HOCs which is why we use `toBeTruthy` to only check that some instance of the
 *   expected element attr exists
 */

export default (TestComponent, enzymeOptions = {}) => {
  /*
   * Test that any html tag or component can be passed through the 'as' prop to
   * override the parent element the component renders
   */
  test('should render as specified html element or component', () => {
    // Create a component to validate that the TestComponent returns.
    const TestAs = () => <section />
    const wrapper = mount(<TestComponent as={TestAs} />, enzymeOptions)
    expect(wrapper.find(TestAs).length).toBeTruthy()
  })

  /*
   * All components should spread passed props on the parent container for max
   * customization
   */
  test('should pass through classnames and attributes', () => {
    const wrapper = mount(
      <TestComponent className="test-custom" data-test="test-custom" />,
      enzymeOptions,
    )
    expect(wrapper.find('.test-custom').length).toBeTruthy()
    expect(wrapper.find(dt('test-custom')).length).toBeTruthy()
  })

  // TODO: should render passed children
}
