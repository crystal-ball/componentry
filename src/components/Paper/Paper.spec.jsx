import { render, screen } from '@testing-library/react'
import { describe, it, expect } from 'vitest'

import { elementTests } from '../../test/element-tests'
import { Paper } from './Paper'

describe('<Paper />', () => {
  elementTests(Paper)
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Paper /> snapshots', () => {
  it('renders correctly', () => {
    render(<Paper data-testid='paper'>Paper content</Paper>)

    expect(screen.getByTestId('paper')).toMatchSnapshot()
  })
})
