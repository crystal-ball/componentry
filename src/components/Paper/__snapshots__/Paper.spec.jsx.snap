// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Paper /> snapshots > renders correctly 1`] = `
<div
  class="C9Y-Paper-base C9Y-Paper-flat"
  data-testid="paper"
>
  Paper content
</div>
`;
